### PR TITLE
Merge to master from conf branch for TACHYON-8 work on new Tachyon configuration

### PR DESCRIPTION
**Pull request by hsaputra**
This is a large PR that contains some changes that are need to complete TACHYON-8 [1] work for new Tachyon configuration after rebase from conf branch:
-) TACHYON-193 - Add new file based Tachyon configuration propertie
-) TACHYON-194 - Add TachyonConf Java class as the only Tachyon runtime config properties
-) TACHYON-198 - Remove existing *Conf classes to be replaced by access to TachyonConf

Missing tasks that should be done in master branch:
-) TACHYON-18 - Refactor tachyon-env.sh to only set Java env variables
-) TACHYON-204 - Move tachyon.version from version.properties file to tachyon-default.properties

We want to merge this quickly before gaps from master and conf branch getting wider.

[1] https://tachyon.atlassian.net/browse/TACHYON-8

### CODE REVIEW

**Comment by hsaputra** [comment]
FYI @haoyuan

**Comment by hsaputra** [comment]
Would love to have this merge to master ASAP to allow some time for updates and follow up changes to complete the new configuration feature.

**Comment by hsaputra** [comment]
I will rebase from master to this PR but please review and merge as soon as possible since it is hard to maintain the differences with more patches coming to master.

Ping @haoyuan

**Comment by hsaputra** [comment]
Update after rebase from master

**Comment by hsaputra** [comment]
Ping @haoyuan for review

File: core/src/main/java/tachyon/Users.java [hunk 1]
```diff
@@ -153,7 +156,8 @@ public synchronized long removeUser(long userId) {
       folder = getUserUfsTempFolder(userId);
       sb.append(" Also remove users underfs folder ").append(folder);
       try {
-        UnderFileSystem.get(CommonConf.get().UNDERFS_ADDRESS).delete(folder, true);
+        String ufsAddress = mTachyonConf.get(Constants.UNDERFS_ADDRESS, "/underfs");
```
Comment (uronce-cc) [inline core/src/main/java/tachyon/Users.java]: @hsaputra What about maintaining all default configuration values as Constants.DEFAULT_XXX?

File: core/src/main/java/tachyon/client/BlockOutStream.java [hunk 2]
```diff
@@ -58,10 +60,12 @@
    * @param file the file the block belongs to
    * @param opType the OutStream's write type
    * @param blockIndex the index of the block in the file
+   * @param tachyonConf the TachyonConf instance for this file output stream.
    * @throws IOException
    */
-  BlockOutStream(TachyonFile file, WriteType opType, int blockIndex) throws IOException {
-    super(file, opType);
+  BlockOutStream(TachyonFile file, WriteType opType, int blockIndex, TachyonConf tachyonConf)
```
Comment (uronce-cc) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: @hsaputra From your detailed design doc, I understand that you want different clients to have different configurations, but I didn't get it when I see so many modifications to function definition or constructors just to add TachyonConf as a new parameter, if it is there just for use of its getInt, getList, etc. methods, maybe we can make these methods static? I may misunderstand your design decision, please feel free to point out.

Comment (hsaputra) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: I am not sure how we could make the methods static? The TachyonConf need to be passed to either constructor or the method call to make sure the same instance is being used.

File: core/src/main/java/tachyon/Users.java (hunk 1 above)
Comment (hsaputra) [inline core/src/main/java/tachyon/Users.java]: Yep, I already add most of them to Constants class, may have missed this one.

File: core/src/main/java/tachyon/client/BlockOutStream.java (hunk 2 above)
Comment (uronce-cc) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: In spark, client can `new SparkConf` and set it to `SparkContext`. 

In tachyon, we can `new TachyonConf` and set it to `TachyonFS`, so that `TachyonFS` serves as something like `TachyonContext`. 

What do you think about this?

Comment (hsaputra) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: As I mentioned before, we need something like SparkContext to hang the TachyonConf. 
I think we could address this in next iteration.

I dont think hang it in TachyonFS is the right solution. Especially when we add modularization of Tachyon then master and workers component may not have access to the  TachyonFS.

Comment (uronce-cc) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: Got it, will be great to see `TachyonContext` in next iteration.

Comment (hsaputra) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: Same here =) 

I think we could hang a lot of stuff in there and not just the configuration instance.

**Comment by hsaputra** [comment]
Reopen with merge from master

File: core/src/main/java/tachyon/client/BlockOutStream.java (hunk 2 above)
Comment (dcapwell) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: How about any changes to the design is anther PR? The idea of a context has the same issue of passing it around.  The only two ways I can see that working is static state that stores the info (which is the world we area trying to move away from) and thread local which is it's own breast that makes the logic more complex.

Really don't think that it's a good idea atm.  Would love to see a simpler design using the idea, but don't think it's a good idea to see in this PR.  @uronce-cc would you be willing to take in the design for that after config goes in?

Comment (uronce-cc) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: @dcapwell Still thinking about the design too, I'll look into spark's design, once I have any idea, I'll talk in jira. Hope to collaborate with @hsaputra

**Comment by hsaputra** [comment]
Ping @haoyuan 

I just merge from master and fix unit tests. Please review this one before pulling others PRs.

The more delay the more painful it is to adjust with new config system.

**Comment by hsaputra** [comment]
Close it for a while to run examples

**Comment by hsaputra** [comment]
Re-open to run the tests

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by hsaputra** [comment]
Weird, it is not repro in my box =(

File: core/src/main/java/tachyon/UserInfo.java [hunk 3]
```diff
@@ -17,7 +17,7 @@
 
 import com.google.common.base.Preconditions;
 
-import tachyon.conf.WorkerConf;
+import tachyon.conf.TachyonConf;
```
Comment (calvinjia) [inline core/src/main/java/tachyon/UserInfo.java]: This appears to be unused currently, is this just for consistency?

File: core/src/main/java/tachyon/Users.java [hunk 4]
```diff
@@ -42,11 +42,12 @@
   private final String mUserUnderFSFolder;
   /** Map from UserId to {@link tachyon.UserInfo} object **/
   private final Map<Long, UserInfo> mUsers;
+  private final TachyonConf mTachyonConf;
 
-  public Users(final String userUfsFolder) {
-
+  public Users( String userUfsFolder, TachyonConf tachyonConf) {
```
Comment (calvinjia) [inline core/src/main/java/tachyon/Users.java]: Extra space. Also, what's the current style guide on final arguments?

Comment (haoyuan) [inline core/src/main/java/tachyon/Users.java]: @hsaputra Could you please also update the style check file to catch this? Thanks.

File: core/src/main/java/tachyon/conf/TachyonConf.java [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+    defaultProps.setProperty(Constants.MASTER_SERVER_THREADS,
+        String.valueOf(2 * Runtime.getRuntime().availableProcessors()));
+
+    InputStream defaultInputStream =
+        TachyonConf.class.getClassLoader().getResourceAsStream(DEFAULT_PROPERTIES);
+    if (defaultInputStream != null) {
+      try {
+        defaultProps.load(defaultInputStream);
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to load default Tachyon properties file.");
```
Comment (calvinjia) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: If there the default properties file is not present, we ignore it, but if it is present but unreadable Tachyon should abort? It seems like these two behaviors are conflicting?

File: core/src/main/java/tachyon/conf/TachyonConf.java [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+
+    // Load site specific properties file
+    Properties siteProps = new Properties();
+    InputStream siteInputStream =
+        TachyonConf.class.getClassLoader().getResourceAsStream(SITE_PROPERTIES);
+    if (siteInputStream != null) {
+      try {
+        siteProps.load(siteInputStream);
+      } catch (IOException e) {
+        LOG.info("Unable to load site Tachyon configuration file.", e);
```
Comment (calvinjia) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: It might be slightly confusing if an info log dumps a stack trace, but I'm not sure if there's a better way (maybe warn level? since the file is present just failed to be read).

File: core/src/main/java/tachyon/conf/TachyonConf.java [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    return updated;
+  }
+
+  public int getInt(String key, final int defaultValue) {
+    if (mProperties.containsKey(key)) {
+      String rawValue = mProperties.getProperty(key);
+      try {
+        return Integer.parseInt(lookup(rawValue));
+      } catch (NumberFormatException e) {
+        LOG.warn("Configuration cannot evaluate key " + key + " as integer.");
```
Comment (calvinjia) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: What's the convention for using {} replacement? In some parts of the code its used even if there is only 1 variable to be populated. Also, perhaps the value or the exception might also be helpful in this case?

**Comment by calvinjia** [comment]
TachyonConf is not intended to be a dynamic configuration, right? For example if the user wanted to update a property, a restart of the cluster (or just workers/masters) would be required (though a dynamic configuration seems to be very useful in practice but more complex). It seems like not all TachyonConf objects are instantiated at the same time, which means different configurations could be in use at the same time?

Also, perhaps instead of using hard coded values for defaults, add all the defaults that would normally be in tachyon-default properties to Constants?

File: core/src/main/java/tachyon/Users.java (hunk 4 above)
Comment (hsaputra) [inline core/src/main/java/tachyon/Users.java]: I will update style check in separate PR. This PR already huge.

Comment (haoyuan) [inline core/src/main/java/tachyon/Users.java]: Agree.

File: core/src/main/java/tachyon/conf/TachyonConf.java (hunk 5 above)
Comment (hsaputra) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: Lets make it consistent by barf on both cases.

File: core/src/main/java/tachyon/conf/TachyonConf.java (hunk 6 above)
Comment (hsaputra) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: It should be fine. This is site override so the non existence or fail to load should be ok.

File: core/src/main/java/tachyon/conf/TachyonConf.java (hunk 7 above)
Comment (hsaputra) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: You mean about logging text? There is no policy right now but should be easy to standardize for later.

**Comment by hsaputra** [comment]
@calvinjia The TachyonConf is not dynamic configuration. The instantiation of the conf is not at the same time and so could be different instance but since we are using file based properties so the out come should be the same.

As for hard-coded default values some of them already in Constants class. I think this could address in next PRs as continuous improvement.

File: core/src/main/java/tachyon/Users.java (hunk 4 above)
Comment (hsaputra) [inline core/src/main/java/tachyon/Users.java]: JIRA: https://tachyon.atlassian.net/browse/TACHYON-278

File: core/src/main/java/tachyon/conf/TachyonConf.java (hunk 7 above)
Comment (hsaputra) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: Usually, any logging that run in normal operation and lower than info should use {} to safe some String materialization.

**Comment by hsaputra** [comment]
@haoyuan Just send update to address comments.

**Comment by calvinjia** [comment]
@hsaputra Got it, so we assume the configuration file is immutable. In practice I am not sure if that is always the case (for example, new conf is pushed to machines of a running cluster before rolling restart), but I think its reasonable at the moment since the difference in configuration values should not cause a huge problem.

File: core/src/main/java/tachyon/conf/TachyonConf.java (hunk 6 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: I think nonexistence is fine, but fail to load usually means the user has some properties to override and failing to load the conf means those properties will not be changed?

**Comment by hsaputra** [comment]
@calvinjia this would be great addition. I did similar thing before and definitely next step in TachyonConf progression.

File: core/src/main/java/tachyon/conf/TachyonConf.java (hunk 6 above)
Comment (hsaputra) [inline core/src/main/java/tachyon/conf/TachyonConf.java]: Ah ok, I guess your proposal to move it up warn to make sure ops could figure out what went wrong is better.

**Comment by hsaputra** [comment]
Resolve comments.

**Comment by hsaputra** [comment]
Ping @haoyuan , all comments are resolved.
Since now not many patches coming it is good time to commit this to master to absorb the changes needed by other PRs before another wave of patches coming in.

File: core/src/main/java/tachyon/client/BlockOutStream.java [hunk 8]
```diff
@@ -59,22 +61,26 @@
...
    * @param blockIndex the index of the block in the file
-   * @param initialBytes the initial size bytes that will be allocated to the block
+   * @param initiallBytes the initial size bytes that will be allocated to the block
+   * @param tachyonConf the TachyonConf instance for this file output stream.
    * @throws IOException
    */
-  BlockOutStream(TachyonFile file, WriteType opType, int blockIndex, long initialBytes)
-      throws IOException {
-    super(file, opType);
+  BlockOutStream(TachyonFile file, WriteType opType, int blockIndex, long initiallBytes,
```
Comment (calvinjia) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: Why was this renamed to initiallBytes?

Comment (hsaputra) [inline core/src/main/java/tachyon/client/BlockOutStream.java]: good catch, probably typo due to so many merge conflict fixes. Will update when it is time to merge.

File: core/src/main/java/tachyon/client/TachyonFS.java [hunk 9]
```diff
@@ -99,18 +102,23 @@ public static synchronized TachyonFS get(final TachyonURI tachyonURI) throws IOE
...
-  public static synchronized TachyonFS get(String masterHost, int masterPort, boolean zookeeperMode)
-      throws IOException {
-    return new TachyonFS(new InetSocketAddress(masterHost, masterPort), zookeeperMode);
+  public static synchronized TachyonFS get(String masterHost, int masterPort,
+      boolean zookeeperMode, TachyonConf tachyonConf) throws IOException {
+    if (tachyonConf != null) {
+      tachyonConf.set(Constants.USE_ZOOKEEPER, Boolean.toString(zookeeperMode));
+      tachyonConf.set(Constants.MASTER_HOSTNAME, masterHost);
+      tachyonConf.set(Constants.MASTER_PORT, Integer.toString(masterPort));
+    }
```
Comment (calvinjia) [inline core/src/main/java/tachyon/client/TachyonFS.java]: Do we set these explicitly because they are not required in the client configuration, I thought set was more of a tool for testing purposes?

Also, is it okay if tachyonConf is null? I think we directly try to access the zookeeper mode later.

Comment (hsaputra) [inline core/src/main/java/tachyon/client/TachyonFS.java]: The set method could be used to modify the TachyonConf state. As for this case, yes it should have been set from client or caller.

I was trying to keep all semantic before to make the changes easy to track. I think now I could change TachyonFS#get to have 2 variance: one is the one that accept just master location, for backward compatibility, and second is the one take TachyonConf. This mean no explicit set should be done here.

Comment (haoyuan) [inline core/src/main/java/tachyon/client/TachyonFS.java]: Since this updated method is not backward compatible anyway, why not just have the new API?

Comment (hsaputra) [inline core/src/main/java/tachyon/client/TachyonFS.java]: Agree! As @calvinjia had recommended we should introduce new API that just take TachyonConf so the internal calls (ie: Tachyon code) will use it instead.

**Comment by haoyuan** [comment]
@hsaputra Thanks for the PR. I have done a pass. Please address the left comments and it should be good to go.

**Comment by hsaputra** [comment]
Thanks so much @haoyuan and all. Will address the rest of comments

**Comment by hsaputra** [comment]
Updated the PR to resolve conflicts and address comments.

File: core/src/main/java/tachyon/client/TachyonFS.java [hunk 10]
```diff
@@ -89,28 +93,49 @@ public static synchronized TachyonFS get(final TachyonURI tachyonURI) throws IOE
         throw new IOException("Invalid Tachyon URI: " + tachyonURI + ". Use " + Constants.HEADER
             + "host:port/ ," + Constants.HEADER_FT + "host:port/");
       }
-      return new TachyonFS(tachyonURI);
+
+      boolean useZookeeper = scheme.equals(Constants.SCHEME_FT);
+      tachyonConf.set(Constants.USE_ZOOKEEPER, Boolean.toString(useZookeeper));
+      tachyonConf.set(Constants.MASTER_HOSTNAME, tachyonURI.getHost());
+      tachyonConf.set(Constants.MASTER_PORT, Integer.toString(tachyonURI.getPort()));
```
Comment (calvinjia) [inline core/src/main/java/tachyon/client/TachyonFS.java]: Are the sets just for the case when the user does not update the new tachyon properties files?

Comment (hsaputra) [inline core/src/main/java/tachyon/client/TachyonFS.java]: That is correct. 

The next step is to actually remove the TachyonConf input argument and let the file based properties file as the main source of truth for customization.
I would love to have the new PR merged to master first so we could add iterative changes and improvements.
It is super hard to maintain this branch with new developments coming in.

**Comment by calvinjia** [comment]
+1 LGTM!

**Comment by hsaputra** [comment]
Thanks for the time to review the changes @calvinjia !

@haoyuan, it is all yours

**Comment by haoyuan** [comment]
Thanks!

**Comment by hsaputra** [comment]
Thanks @haoyuan 

CC @dcapwell, @rootfs, @calvinjia, @uronce-cc, @shimingfei, @GraceH, @childsb 

Please give it a spin and let's make it better.
